File: freepascal/parametrized_SQL_statement/paramsqlstat.pas
program Parametrized_SQL_Statement;
uses
  sqlite3, sysutils;

const
  DB_FILE : PChar = 'test.db';

var
  DB                :Psqlite3;
  ResultCode        :Integer;
  SQL               :PChar;
  CompiledStatement :Psqlite3_stmt;

Procedure InsertValues(DB :Psqlite3);
var
  i             :integer;
  ResultCode    :integer;
  SQLStatements :array [1..4] of PChar;
begin
  SQLStatements[1] := 'insert into players (name, score, active, jerseyNum) values (''Ben Roethlisberger'', 94.1, 1, 7);';
  SQLStatements[2] := 'insert into players (name, score, active, jerseyNum) values (''Alex Smith'', 85.3, 1, 11);';
  SQLStatements[3] := 'insert into players (name, score, active, jerseyNum) values (''Payton Manning'', 96.5, 0, 18);';
  SQLStatements[4] := 'insert into players (name, score, active, jerseyNum) values (''John Doe'', 15, 0, 99);';

  for i:=1 to 4 do
  begin
    ResultCode := sqlite3_exec(DB, SQLStatements[i], nil, nil, nil);
    if ResultCode <> SQLITE_OK then
    begin
      writeln(format('Error #%d: %s', [ResultCode, sqlite3_errmsg(db)]));
      sqlite3_close(DB);
      halt(ResultCode);
    end;
  end;
end;

begin
  // Open the database.
  ResultCode := sqlite3_open(DB_FILE, @DB);
  if ResultCode <> SQLITE_OK then
  begin
    writeln(format('%d: %s', [ResultCode, sqlite3_errmsg(db)]));
    halt(1);
  end;

  // Create the players table in the database.
  SQL := 'create table players(id integer primary key asc, name text, score real, active integer, jerseyNum integer);';
  ResultCode := sqlite3_exec(DB, SQL, nil, nil, nil);
  if ResultCode <> SQLITE_OK then
  begin
    writeln(format('Error #%d: %s', [ResultCode, sqlite3_errmsg(db)]));
    sqlite3_close(DB);
    halt(ResultCode);
  end;

  // Insert some values into the players table.
  InsertValues(DB);

  // Update player #99.
  // TODO

  sqlite3_close(db);
end.

File: c/parametrized_SQL_statement/paramsqlstat.c
#include <stdio.h>
#include <stdlib.h>
#include <sqlite3.h>

static const char* db_file = ":memory:"; // Create an in-memory database.  

void check_error(int result_code, sqlite3 *db);
int select_callback(void* data, int column_count, char** columns, char** column_names);

int main(void) {
    sqlite3 *db;
    int result_code;
    char *sql;
    char *insert_statements[4];
    sqlite3_stmt *compiled_statement;
    int i;
    
    // Open the database.
    result_code = sqlite3_open(db_file, &db);
    check_error(result_code, db);    
    
    // Create the players table in the database.
    sql = "create table players("
          "id integer primary key asc, "
          "name text, "
          "score real, "
          "active integer, " // Store the bool value as integer (see https://sqlite.org/datatype3.html chapter 2.1).
          "jerseyNum integer);";
    result_code = sqlite3_exec(db, sql, NULL, NULL, NULL);
    check_error(result_code, db);    
        
    // Insert some values into the players table.    
    insert_statements[0] = "insert into players (name, score, active, jerseyNum) "
                                        "values ('Roethlisberger, Ben', 94.1, 1, 7);";
    insert_statements[1] = "insert into players (name, score, active, jerseyNum) "
                                        "values ('Smith, Alex', 85.3, 1, 11);";
    insert_statements[2] = "insert into players (name, score, active, jerseyNum) "
                                        "values ('Manning, Payton', 96.5, 0, 18);";
    insert_statements[3] = "insert into players (name, score, active, jerseyNum) "
                                        "values ('Doe, John', 15, 0, 99);";
    
    for (i=0; i<4; i++) {
        result_code = sqlite3_exec(db, insert_statements[i], NULL, NULL, NULL);        
        check_error(result_code, db);    
    }
        
    // Display the contents of the players table.
    printf("Before update:\n");
    sql = "select * from players;";
    result_code = sqlite3_exec(db, sql, select_callback, NULL, NULL);
    check_error(result_code, db);    
    
    // Close the database connection.
    sqlite3_close(db);        
        
    return EXIT_SUCCESS;
}

void check_error(int result_code, sqlite3 *db) {
    if (result_code != SQLITE_OK) {
        printf("Error #%d: %s\n", result_code, sqlite3_errmsg(db));
        sqlite3_close(db);
        exit(result_code);
    }
}

int select_callback(void* data, int column_count, char** columns, char** column_names) {
    int i;
    
    for (i=0; i<column_count; i++) {
        printf(columns[i]);
        if (i < column_count-1) printf(" | ");
    }
    printf("\n");
}

